src/XR/features/WebXRControllerMovement.ts

```diff
@@ -0,0 +1,547 @@
+    /**
+     * For registering movement to specific hand only.  Useful if your app has a "main hand" and "off hand" for determining the functionality of a controller.
+     */
+    handedness?: XRHandedness
+    /**
+     * For main component only (useful for buttons and may not trigger axis changes)
+     */
+    mainComponentOnly?: boolean
+    /**
+     * Additional predicate to apply to controllers to restrict a handler being added.
+     */
+    componentSelectionPredicate?: (xrController: WebXRInputSource) => Nullable<WebXRControllerComponent>
+} & ({
+    /**
+     * called when the axis changes
+     */
+    axisChangedHandler: (axes: IWebXRMotionControllerAxesValue, movementState: WebXRControllerMovementState, featureContext: WebXRControllerMovementFeatureContext, xrInput: WebXRInput) => void
+    // buttonChangedhandler?: never
+} | {
+    /**
+     * Called when the button state changes
+     */
+    buttonChangedhandler: (pressed: IWebXRMotionControllerComponentChangesValues<boolean>, movementState: WebXRControllerMovementState, featureContext: WebXRControllerMovementFeatureContext, xrInput: WebXRInput) => void
+    // axisChangedHandler?: never
+});
+
+type RegisteredComponent = {
+    registrationConfiguration: WebXRControllerMovementRegistrationConfiguration
+    component: WebXRControllerComponent
+    onAxisChangedObserver?: Nullable<Observer<IWebXRMotionControllerAxesValue>>
+    onButtonChangedObserver?: Nullable<Observer<WebXRControllerComponent>>
+};
+
+/**
+ * This is a movement feature to be used with WebXR-enabled motion controllers.
+ * When enabled and attached, the feature will allow a user to move around and rotate in the scene using
+ * the input of the attached controllers.
+ */
+export class WebXRControllerMovement extends WebXRAbstractFeature {
+    private _controllers: {
+        [controllerUniqueId: string]: {
+            xrController: WebXRInputSource;
+            registeredComponents: RegisteredComponent[];
+        };
+    } = {};
+
+    private _currentRegistrationConfigurations: WebXRControllerMovementRegistrationConfiguration[] = [];
+    // Feature configuration is syncronized - this is passed to all handlers (reduce GC pressure).
+    private _featureContext: WebXRControllerMovementFeatureContext;
+    // forward direction for movement, which may differ from viewer pose.
+    private _movementDirection: Nullable<Quaternion> = null;
+    private _movementState: WebXRControllerMovementState;
+    private _xrInput: WebXRInput;
+
+    // unused
+    private _rotationMatrix: Matrix = Matrix.Identity();
+    private _translationDirection: Vector3 = new Vector3();
+    private _movementTranslation: Vector3 = new Vector3();
+
+    /**
+     * The module's name
+     */
+    public static readonly Name = WebXRFeatureName.MOVEMENT;
+
+    /**
+     * Standard controller configurations.
+     */
+    public static readonly REGISTRATIONS: Record<string, WebXRControllerMovementRegistrationConfiguration[]> = {

```

Comment:
I find `{[key:string]: WebXRControllerMovementRegistrationConfiguration[]}` better, especially since we don't use Record in the framework. Is there a difference between the two?

Reply:
agreed.  i think it's better too, more natural for strings.